### ISSUE TITLE
Add atan2 trigonometry function

### ISSUE THREAD

**Issue by jphalip** [tensorflow/tensorflow#6095]
Tensorflow has a number of built-in trigonometry functions like `atan` but it doesn't have `atan2` yet.

This can be achieved with some custom code as pointed in this comment: https://github.com/tensorflow/tensorflow/issues/3624#issuecomment-242085609.

There's actually a small mistake on the first line, where `np.pi` should not be added. The corrected code is presented below:

```python
def atan2(y, x):
    angle = tf.select(tf.greater(x,0.0), tf.atan(y/x), tf.zeros_like(x))
    angle = tf.select(tf.logical_and(tf.less(x,0.0),  tf.greater_equal(y,0.0)), tf.atan(y/x) + np.pi, angle)
    angle = tf.select(tf.logical_and(tf.less(x,0.0),  tf.less(y,0.0)), tf.atan(y/x) - np.pi, angle)
    angle = tf.select(tf.logical_and(tf.equal(x,0.0), tf.greater(y,0.0)), 0.5*np.pi * tf.ones_like(x), angle)
    angle = tf.select(tf.logical_and(tf.equal(x,0.0), tf.less(y,0.0)), -0.5*np.pi * tf.ones_like(x), angle)
    angle = tf.select(tf.logical_and(tf.equal(x,0.0), tf.equal(y,0.0)), np.nan * tf.zeros_like(x), angle)
    return angle
```

While this custom code performs reasonably well, it'd be both faster and more convenient if Tensorflow had it built-in like the other existing functions.

**Comment by michaelisard** [comment]
Thanks, please feel free to submit a PR and reference this issue.

**Comment by tesfaldet** [comment]
Is there any way to convert the range to between 0 and 2*pi? I'm trying to mimic opencv's cartToPolar method this way, however, I'm having trouble finding a way to modify elements in a tensor (on a placeholder, so I can't use scatter update) based off of a condition i.e. where angle < 0, add 2*pi.

**Comment by tesfaldet** [comment]
For what it's worth, I managed to implement two versions of atan2. One that follows the opencv style (returns degrees), and another that uses tensorflow's atan function (returns rads).

`def atan2(y, x):
    angle = tf.select(tf.greater(x, 0.0), tf.atan(y / x), tf.zeros_like(x))
    angle = tf.select(tf.greater(y, 0.0), 0.5 * np.pi - tf.atan(x / y), angle)
    angle = tf.select(tf.less(y, 0.0), -0.5 * np.pi - tf.atan(x / y), angle)
    angle = tf.select(tf.less(x, 0.0), tf.atan(y / x) + np.pi, angle)
    angle = tf.select(tf.logical_and(tf.equal(x, 0.0), tf.equal(y, 0.0)),
                      np.nan * tf.zeros_like(x), angle)

    indices = tf.where(tf.less(angle, 0.0))
    updated_values = tf.gather_nd(angle, indices) + (2 * np.pi)
    update = tf.SparseTensor(indices, updated_values, angle.get_shape())
    update_dense = tf.sparse_tensor_to_dense(update)

    return angle + update_dense


def atan2_ocv(y, x):
    # constants
    DBL_EPSILON = 2.2204460492503131e-16
    atan2_p1 = 0.9997878412794807 * (180 / np.pi)
    atan2_p3 = -0.3258083974640975 * (180 / np.pi)
    atan2_p5 = 0.1555786518463281 * (180 / np.pi)
    atan2_p7 = -0.04432655554792128 * (180 / np.pi)

    ax, ay = tf.abs(x), tf.abs(y)
    c = tf.select(tf.greater_equal(ax, ay), tf.div(ay, ax + DBL_EPSILON),
                  tf.div(ax, ay + DBL_EPSILON))
    c2 = tf.square(c)
    angle = (((atan2_p7 * c2 + atan2_p5) * c2 + atan2_p3) * c2 + atan2_p1) * c
    angle = tf.select(tf.greater_equal(ax, ay), angle, 90.0 - angle)
    angle = tf.select(tf.less(x, 0.0), 180.0 - angle, angle)
    angle = tf.select(tf.less(y, 0.0), 360.0 - angle, angle)
    return angle`

I'm not sure if I should add or subtract np.pi when x < 0.0 for atan2, since it seems to be a platform dependent thing, so I left it at adding np.pi. I validated the output of my atan2_ocv by using it for visualizing optical flow, and comparing that with opencv's flow visualization implementation which uses cv2.cartToPolar (which in turn uses its own implementation of atan2).

**Comment by tesfaldet** [comment]
Oh, and if anyone's coming here to look for an atan2 function for their own flow visualization function, here's what I made (I know this isn't the point of this issue!):

```python
def normalize(tensor, a=0, b=1):
    return tf.div(tf.mul(tf.sub(tensor, tf.reduce_min(tensor)), b - a),
                  tf.sub(tf.reduce_max(tensor), tf.reduce_min(tensor)))


def cart_to_polar_ocv(x, y, angle_in_degrees=False):
    v = tf.sqrt(tf.add(tf.square(x), tf.square(y)))
    ang = atan2_ocv(y, x)
    scale = 1 if angle_in_degrees else np.pi / 180
    return v, tf.mul(ang, scale)


def cart_to_polar(x, y, angle_in_degrees=False):
    v = tf.sqrt(tf.add(tf.square(x), tf.square(y)))
    ang = atan2(y, x)
    scale = 180 / np.pi if angle_in_degrees else 1
    return v, tf.mul(ang, scale)


def draw_flow(flow):
    fx, fy = flow[:, :, :, 0], flow[:, :, :, 1]
    v, ang = cart_to_polar_ocv(fx, fy)

    h = normalize(tf.mul(ang, 180 / np.pi))
    s = tf.ones_like(h)
    v = normalize(v)

    hsv = tf.pack([h, s, v], 3)
    rgb = tf.image.hsv_to_rgb(hsv) * 255

    return tf.cast(rgb, tf.uint8)
```

**Comment by mjlm** [comment]
As far as I can tell, the implementation mentioned in the first post will suffer from division-by-zero if x or y contain any zeros, which leads to NaN gradients. This fixes the issue for me:

```python
def atan2(y, x, epsilon=1.0e-12):
    # Add a small number to all zeros, to avoid division by zero:
    x = tf.select(tf.equal(x, 0.0), x+epsilon, x)
    y = tf.select(tf.equal(y, 0.0), y+epsilon, y)

    angle = tf.select(tf.greater(x,0.0), tf.atan(y/x), tf.zeros_like(x))
    angle = tf.select(tf.logical_and(tf.less(x,0.0),  tf.greater_equal(y,0.0)), tf.atan(y/x) + np.pi, angle)
    angle = tf.select(tf.logical_and(tf.less(x,0.0),  tf.less(y,0.0)), tf.atan(y/x) - np.pi, angle)
    angle = tf.select(tf.logical_and(tf.equal(x,0.0), tf.greater(y,0.0)), 0.5*np.pi * tf.ones_like(x), angle)
    angle = tf.select(tf.logical_and(tf.equal(x,0.0), tf.less(y,0.0)), -0.5*np.pi * tf.ones_like(x), angle)
    angle = tf.select(tf.logical_and(tf.equal(x,0.0), tf.equal(y,0.0)), tf.zeros_like(x), angle)
    return angle
```

**Comment by alquraishi** [comment]
Any updates on adding builtin atan2 support?

**Comment by asos-ben** [comment]
I think that select is deprecated in version 1.0, so this is now

`def atan2(x, y, epsilon=1.0e-12):
        """
        A hack until the tf developers implement a function that can find the angle from an x and y co-ordinate.
        :param x: 
        :param epsilon: 
        :return: 
        """
        # Add a small number to all zeros, to avoid division by zero:
        x = tf.where(tf.equal(x, 0.0), x+epsilon, x)
        y = tf.where(tf.equal(y, 0.0), y+epsilon, y)
    
        angle = tf.where(tf.greater(x,0.0), tf.atan(y/x), tf.zeros_like(x))
        angle = tf.where(tf.logical_and(tf.less(x,0.0),  tf.greater_equal(y,0.0)), tf.atan(y/x) + np.pi, angle)
        angle = tf.where(tf.logical_and(tf.less(x,0.0),  tf.less(y,0.0)), tf.atan(y/x) - np.pi, angle)
        angle = tf.where(tf.logical_and(tf.equal(x,0.0), tf.greater(y,0.0)), 0.5*np.pi * tf.ones_like(x), angle)
        angle = tf.where(tf.logical_and(tf.equal(x,0.0), tf.less(y,0.0)), -0.5*np.pi * tf.ones_like(x), angle)
        angle = tf.where(tf.logical_and(tf.equal(x,0.0), tf.equal(y,0.0)), tf.zeros_like(x), angle)
        return angle`

**Comment by girving** [comment]
Fixed by https://github.com/tensorflow/tensorflow/pull/9276.
